pkg/blockstorage/vmware/vmware.go

```diff
@@ -253,7 +255,8 @@ func (p *FcdProvider) SnapshotCreate(ctx context.Context, volume blockstorage.Vo
 	var res types.AnyType
 	err := wait.PollImmediate(time.Second, defaultRetryLimit, func() (bool, error) {
 		log.Debug().Print("CreateSnapshot", field.M{"VolumeID": volume.ID})
-		task, lerr := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), noDescription)
+		description := generateDescription(tags)
+		task, lerr := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), description)
```

Comment:
Thanks, @carlbraganza 
Yes, I'm currently working on this part. Here is a commit, I'm going to test it today
https://github.com/kanisterio/kanister/pull/1555/commits/8c7cc05000f94fdd8ba2e5a8415b7e7b1b472d15

Reply:
Instead of adding multiple checks to see if the snapshot is truly created in `wait.PollImmediate`, can we split it up then snapshot creation and error handling parts? Feels like separating them makes the error handling part easier to reason about, and more testable.

So something like this:
```diff
diff --git a/pkg/blockstorage/vmware/vmware.go b/pkg/blockstorage/vmware/vmware.go
index 9952c040..4beee14b 100644
--- a/pkg/blockstorage/vmware/vmware.go
+++ b/pkg/blockstorage/vmware/vmware.go
@@ -255,48 +255,52 @@ func (p *FcdProvider) SnapshotCreate(ctx context.Context, volume blockstorage.Vo
 	var res types.AnyType
 	description := generateDescription(tags)
 	err := wait.PollImmediate(time.Second, defaultRetryLimit, func() (bool, error) {
-		log.Debug().Print("CreateSnapshot", field.M{"VolumeID": volume.ID})
-		task, lerr := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), description)
-		if lerr != nil {
-			res := p.getCreatedSnapshot(ctx, tags, volume.ID, time.Now().Add(-1*vmWareTimeout))
-			if res != nil {
-				log.Error().WithError(lerr).Print("Wait failed but snapshot was created")
-				return true, nil
+		createErr := func() error {
+			log.Debug().Print("CreateSnapshot", field.M{"VolumeID": volume.ID})
+			task, lerr := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), description)
+			if lerr != nil {
+				return errors.Wrap(lerr, "CreateSnapshot task creation failure")
 			}
-			return false, errors.Wrap(lerr, "CreateSnapshot task creation failure")
-		}
-		log.Debug().Print("Started CreateSnapshot task", field.M{"VolumeID": volume.ID})
-		res, lerr = task.Wait(ctx, vmWareTimeout)
-		if lerr != nil {
-			res := p.getCreatedSnapshot(ctx, tags, volume.ID, time.Now().Add(-1*vmWareTimeout))
-			if res != nil {
-				log.Error().WithError(lerr).Print("Wait failed but snapshot was created")
-				return true, nil
+			log.Debug().Print("Started CreateSnapshot task", field.M{"VolumeID": volume.ID})
+			res, lerr = task.Wait(ctx, vmWareTimeout)
+			if lerr != nil {
+				return lerr
 			}
-			// check if snapshot was created
-			if soap.IsVimFault(lerr) {
-				switch soap.ToVimFault(lerr).(type) {
-				case *types.InvalidState:
-					log.Error().WithError(lerr).Print("There is some operation, other than this CreateSnapshot invocation, on the VM attached still being protected by its VM state. Will retry", field.M{"VolumeID": volume.ID})
-					return false, nil
-				case *vslmtypes.VslmSyncFault: // potentially can leak snapshots
-					if vFault, ok := soap.ToVimFault(lerr).(*vslmtypes.VslmSyncFault); ok {
-						log.Error().Print(fmt.Sprintf("VslmSyncFault: %#v", vFault))
-					}
-					if !(govmomiError{lerr}).Matches(reVslmSyncFaultFatal) {
-						log.Error().Print(fmt.Sprintf("CreateSnapshot failed with VslmSyncFault. Will retry: %s", (govmomiError{lerr}).Format()), field.M{"VolumeID": volume.ID})
-						return false, nil
-					}
-					return false, errors.Wrap(lerr, "CreateSnapshot failed with VslmSyncFault. A snapshot may have been created by this failed operation")
-				case *types.NotFound:
-					log.Error().WithError(lerr).Print("CreateSnapshot failed with NotFound error. Will retry", field.M{"VolumeID": volume.ID})
+			return nil
+		}()
+
+		if createErr == nil {
+			return true, nil
+		}
+
+		// it's possible that snapshot was created despite of SOAP errors
+		res := p.getCreatedSnapshot(ctx, tags, volume.ID, time.Now().Add(-1*vmWareTimeout))
+		if res != nil {
+			log.Error().WithError(createErr).Print("snapshot created with errors")
+			return true, nil
+		}
+
+		// snapshot wasn't created, handle the different SOAP errors then retry
+		if soap.IsVimFault(createErr) {
+			switch soap.ToVimFault(createErr).(type) {
+			case *types.InvalidState:
+				log.Error().WithError(createErr).Print("There is some operation, other than this CreateSnapshot invocation, on the VM attached still being protected by its VM state. Will retry", field.M{"VolumeID": volume.ID})
+				return false, nil
+			case *vslmtypes.VslmSyncFault: // potentially can leak snapshots
+				if vFault, ok := soap.ToVimFault(createErr).(*vslmtypes.VslmSyncFault); ok {
+					log.Error().Print(fmt.Sprintf("VslmSyncFault: %#v", vFault))
+				}
+				if !(govmomiError{createErr}).Matches(reVslmSyncFaultFatal) {
+					log.Error().Print(fmt.Sprintf("CreateSnapshot failed with VslmSyncFault. Will retry: %s", (govmomiError{createErr}).Format()), field.M{"VolumeID": volume.ID})
 					return false, nil
 				}
+				return false, errors.Wrap(createErr, "CreateSnapshot failed with VslmSyncFault. A snapshot may have been created by this failed operation")
+			case *types.NotFound:
+				log.Error().WithError(createErr).Print("CreateSnapshot failed with NotFound error. Will retry", field.M{"VolumeID": volume.ID})
+				return false, nil
 			}
-			return false, errors.Wrap(lerr, "Failed to wait on task")
 		}
-		log.Debug().Print("CreateSnapshot task complete", field.M{"VolumeID": volume.ID})
-		return true, nil
+		return false, errors.Wrap(createErr, "Failed to wait on task")
 	})
 	if err != nil {
 		log.Error().WithError(err).Print(fmt.Sprintf("Failed to create snapshot for FCD %s: %s", volume.ID, govmomiError{err}.Format()))
```
Feels like this is also more testable, if we extract them out into separate functions.

Note: just a take-it-or-leave-it suggestion.

Reply:
Yes, I've also noticed that I have a duplicates but I didn't find any suitable solution. Your proposal looks good I'll take it